### ISSUE TITLE
httpd lens fails on backslashes in directive regular expressions

### ISSUE THREAD

**Issue by erebus2** [hercules-team/augeas#307]
Augeas currently throws “syntax errors” when an apache config happens to include regular expressions.

A regular expressions is not a syntax error. In fact, regexes are quite common in apache configs as they are needed for mod_rewrite.

This issue is happening with the httpd lens from augeas v1.4.

See also: https://github.com/letsencrypt/letsencrypt/issues/1202

**Comment by domcleal** [comment]
Could you provide part of a config file that fails please?

**Comment by pde** [comment]
There's a sample in this ticket:

https://github.com/letsencrypt/letsencrypt/issues/1264

**Comment by domcleal** [comment]
@pde I don't see a regular expression there.  The config file in that ticket works fine on Augeas 1.4.0.

**Comment by pde** [comment]
I suspect these may have been the samples we were looking for:

```
#
# Apache/PHP/Drupal settings:
#

# Protect files and directories from prying eyes.
<FilesMatch "\.(engine|inc|info|install|make|module|profile|test|po|sh|.*sql|theme|tpl(\.php)?|xtmpl)(~|\.sw[op]|\.bak|\.orig|\.save)?$|^(\..*|Entries.*|Repository|Root|Tag|Template)$|^#.*#$|\.php(~|\.sw[op]|\.bak|\.orig\.save)$">
  Order allow,deny
</FilesMatch>

# Don't show directory listings for URLs which map to a directory.
Options -Indexes

# Follow symbolic links in this directory.
Options +FollowSymLinks

# Make Drupal handle any 404 errors.
ErrorDocument 404 /index.php

# Set the default handler.
DirectoryIndex index.php index.html index.htm
```

```
# Apache config file for anarcat.wiki.orangeseeds.org
# Automatically generated by ikisite; do not modify directly.

<VirtualHost *:80>
        ServerAdmin root@localhost
        ServerName anarcat.wiki.orangeseeds.org:80


        SuexecUserGroup w-anarcat w-anarcat


        UserDir disabled

        RewriteEngine On
        RewriteRule ^/(.*) http\:\/\/anarc\.at\/$1 [L,R,NE]

        ErrorLog /var/log/ikiwiki-hosting/w-anarcat/error.log
        LogLevel warn
        CustomLog /var/log/ikiwiki-hosting/w-anarcat/access.log combined
</VirtualHost>
```

**Comment by pde** [comment]
Another sample over at https://github.com/letsencrypt/letsencrypt/issues/1598

**Comment by domcleal** [comment]
Yep, the second sample there is failing to parse on master, thanks for the report.

**Comment by pde** [comment]
@domcleal two questions for you:

1. are you saying the first (Drupal .htaccess) example *does* parse on master, so we should be investigating it for some other problem with letsencrypt?

2. please let us know if there's a realistic chance that the augeas team might have a fix for this by early next week.  If so, we'll plan to ship it before our Public Beta launch next Thursday!

**Comment by ahaaje** [comment]
> Another sample over at letsencrypt/letsencrypt#1598

I had a similar case with the letsencrypt client

This failed to parse
```
RewriteEngine on 
RewriteCond %{HTTP_HOST} !^www\.example\.no 
RewriteRule ^(.*)$ http://www.example.no/$1 [R=301,L]
```

Commenting out by putting a hash tag in front of RewriteCond let it continue. It was not necesarry to comment out RewriteRule. 

This was allowed to continue

```
RewriteEngine on 
#RewriteCond %{HTTP_HOST} !^www\.example\.no 
RewriteRule ^(.*)$ http://www.example.no/$1 [R=301,L]
```

This was in one of several VirtualHost example.conf files.

**Comment by domcleal** [comment]
> are you saying the first (Drupal .htaccess) example does parse on master, so we should be investigating it for some other problem with letsencrypt?

Yes, it worked for me.  It's possible some subtle parts were lost when putting it into the comment if the original file isn't working for you, e.g. bug #312.  If that's the case, it might be best to upload the file somewhere.

> please let us know if there's a realistic chance that the augeas team might have a fix for this by early next week. If so, we'll plan to ship it before our Public Beta launch next Thursday!

I'll have a look, but it was a bit fiddly.

> Commenting out by putting a hash tag in front of RewriteCond let it continue. It was not necesarry to comment out RewriteRule.

Thanks, that confirms my quick investigation which shows it's the backslash inside the regexes.  Quoting them (if that's valid in httpd's config) worked around it too.

**Comment by domcleal** [comment]
PR #318 should fix this issue.

**Comment by domcleal** [comment]
Fixed in 155746c.
